source/slang/slang-lower-to-ir.cpp

```diff
@@ -1498,7 +1492,24 @@ struct ValLoweringVisitor : ValVisitor<ValLoweringVisitor, LoweredValInfo, Lower
         auto intrinsicTypeModifier = type->declRef.getDecl()->findModifier<IntrinsicTypeModifier>();
         SLANG_ASSERT(intrinsicTypeModifier);
         IROp op = IROp(intrinsicTypeModifier->irOp);
-        return getBuilder()->getType(op);
+        List<IRInst*> operands;
+        // If there are any substitutions attached to the declRef,
+        // add them as operands of the IR type.
+        for (auto subst = type->declRef.substitutions.substitutions;
+            subst; subst = subst->outer)
```

Comment:
It shouldn't matter for any type we have today, but I think this would end up with a backwards result in the case of nested generics. If we have:

```hlsl
struct Outer<T>
{
    struct Inner<U>
    {
    }
}
```

then the iteration here would produce an IR instruction for `Outer<X>.Inner<Y>` that is of the form `InnerOp(Y, X)`. That is because the `substitutions` linked list for a `DeclRef` always starts with the inner-most substitutions and proceeds outward (so that it can share the outer substitutions with the parent declaration).

I suspect that the official Right Way to handle this would be to have a recursive function on the substitutions, that adds arguments for the outer substitution before adding arguments for the current one.

For now this really doesn't matter, since we have no cases of nested generic types that need to map to intrinsic operations in this fashion.

Reply:
Done.